File: repositories/doctor_repository.go
package repositories

import (
	"context"
	"errors"
	"time"

	"doctor-service/entities"

	"github.com/google/uuid"
	"gorm.io/gorm"
)

type DoctorRepository interface {
	Create(ctx context.Context, doctor *entities.Doctor) error
	FindByEmail(ctx context.Context, email string) (*entities.Doctor, error)
	FindByMobile(ctx context.Context, mobile string) (*entities.Doctor, error)
	FindByIdentifier(ctx context.Context, identifier string) (*entities.Doctor, error)
	FindByPublicID(ctx context.Context, publicID uuid.UUID) (*entities.Doctor, error)
	Update(ctx context.Context, doctor *entities.Doctor) error
	IncrementFailedLogin(ctx context.Context, doctorID uuid.UUID, maxAttempts int, lockDuration time.Duration) (int, bool, error)
	ResetFailedLogin(ctx context.Context, doctorID uuid.UUID) error
	MarkMobileVerified(ctx context.Context, doctorID uuid.UUID) error
}

type GormDoctorRepository struct {
	db *gorm.DB
}

func NewDoctorRepository(db *gorm.DB) DoctorRepository {
	return &GormDoctorRepository{db: db}
}

func (r *GormDoctorRepository) Create(ctx context.Context, doctor *entities.Doctor) error {
	return r.db.WithContext(ctx).Create(doctor).Error
}

func (r *GormDoctorRepository) FindByEmail(ctx context.Context, email string) (*entities.Doctor, error) {
	var doc entities.Doctor
	err := r.db.WithContext(ctx).Where("email = ?", email).First(&doc).Error
	if errors.Is(err, gorm.ErrRecordNotFound) {
		return nil, nil
	}
	return &doc, err
}

func (r *GormDoctorRepository) FindByMobile(ctx context.Context, mobile string) (*entities.Doctor, error) {
	var doc entities.Doctor
	err := r.db.WithContext(ctx).Where("mobile = ?", mobile).First(&doc).Error
	if errors.Is(err, gorm.ErrRecordNotFound) {
		return nil, nil
	}
	return &doc, err
}

func (r *GormDoctorRepository) FindByIdentifier(ctx context.Context, identifier string) (*entities.Doctor, error) {
	var doc entities.Doctor
	err := r.db.WithContext(ctx).Where("email = ? OR mobile = ?", identifier, identifier).First(&doc).Error
	if errors.Is(err, gorm.ErrRecordNotFound) {
		return nil, nil
	}
	return &doc, err
}

func (r *GormDoctorRepository) FindByPublicID(ctx context.Context, publicID uuid.UUID) (*entities.Doctor, error) {
	var doc entities.Doctor
	err := r.db.WithContext(ctx).Where("public_id = ?", publicID).First(&doc).Error
	if errors.Is(err, gorm.ErrRecordNotFound) {
		return nil, nil
	}
	return &doc, err
}

func (r *GormDoctorRepository) Update(ctx context.Context, doctor *entities.Doctor) error {
	return r.db.WithContext(ctx).Save(doctor).Error
}

func (r *GormDoctorRepository) IncrementFailedLogin(ctx context.Context, doctorID uuid.UUID, maxAttempts int, lockDuration time.Duration) (int, bool, error) {
	var doc entities.Doctor
	if err := r.db.WithContext(ctx).First(&doc, "id = ?", doctorID).Error; err != nil {
		return 0, false, err
	}

	doc.FailedLoginAttempts++
	isLocked := false
	if doc.FailedLoginAttempts >= maxAttempts {
		lockedUntil := time.Now().Add(lockDuration)
		doc.AccountLockedUntil = &lockedUntil
		isLocked = true
	}

	err := r.db.WithContext(ctx).Model(&doc).Updates(map[string]interface{}{
		"failed_login_attempts": doc.FailedLoginAttempts,
		"account_locked_until":  doc.AccountLockedUntil,
	}).Error

	return doc.FailedLoginAttempts, isLocked, err
}

func (r *GormDoctorRepository) ResetFailedLogin(ctx context.Context, doctorID uuid.UUID) error {
	now := time.Now()
	return r.db.WithContext(ctx).Model(&entities.Doctor{}).Where("id = ?", doctorID).Updates(map[string]interface{}{
		"failed_login_attempts": 0,
		"account_locked_until":  nil,
		"last_login_at":         &now,
	}).Error
}

func (r *GormDoctorRepository) MarkMobileVerified(ctx context.Context, doctorID uuid.UUID) error {
	return r.db.WithContext(ctx).Model(&entities.Doctor{}).Where("id = ?", doctorID).Update("mobile_verified", true).Error
}

File: backend_api/repositories/doctor_repository.py
import datetime
from typing import Optional, Tuple
from uuid import UUID
from sqlalchemy.orm import Session
from sqlalchemy import or_
from entities.doctor import Doctor

class DoctorRepository:
    def __init__(self, db: Session):
        self.db = db

    def Create(self, doctor: Doctor) -> None:
        self.db.add(doctor)
        self.db.commit()
        self.db.refresh(doctor)

    def FindByEmail(self, email: str) -> Optional[Doctor]:
        return self.db.query(Doctor).filter(Doctor.email == email).first()

    def FindByMobile(self, mobile: str) -> Optional[Doctor]:
        return self.db.query(Doctor).filter(Doctor.mobile == mobile).first()

    def FindByIdentifier(self, identifier: str) -> Optional[Doctor]:
        return self.db.query(Doctor).filter(or_(Doctor.email == identifier, Doctor.mobile == identifier)).first()

    def FindByPublicID(self, public_id: UUID) -> Optional[Doctor]:
        return self.db.query(Doctor).filter(Doctor.public_id == str(public_id)).first()

    def Update(self, doctor: Doctor) -> None:
        self.db.merge(doctor)
        self.db.commit()

    def IncrementFailedLogin(self, doctor_id: UUID, max_attempts: int, lock_duration: datetime.timedelta) -> Tuple[int, bool]:
        doc = self.db.query(Doctor).filter(Doctor.id == str(doctor_id)).first()
        if not doc:
            raise Exception("Doctor not found")
        
        doc.failed_login_attempts += 1
        is_locked = False
        if doc.failed_login_attempts >= max_attempts:
            locked_until = datetime.datetime.now() + lock_duration
            doc.account_locked_until = locked_until
            is_locked = True
            
        self.db.commit()
        self.db.refresh(doc)
        return doc.failed_login_attempts, is_locked

    def ResetFailedLogin(self, doctor_id: UUID) -> None:
        now = datetime.datetime.now()
        self.db.query(Doctor).filter(Doctor.id == str(doctor_id)).update({
            "failed_login_attempts": 0,
            "account_locked_until": None,
            "last_login_at": now
        })
        self.db.commit()

    def MarkMobileVerified(self, doctor_id: UUID) -> None:
        self.db.query(Doctor).filter(Doctor.id == str(doctor_id)).update({"mobile_verified": True})
        self.db.commit()
